# PART - I

In [39]:
import numpy as np 
import pandas as pd 

parsed_df = pd.read_csv("CSV_Files/parsed.csv")

parsed_df.head()

alert  cdi      code                                             detail  \
0   NaN  NaN  37389218  https://earthquake.usgs.gov/fdsnws/event/1/que...   
1   NaN  NaN  37389202  https://earthquake.usgs.gov/fdsnws/event/1/que...   
2   NaN  4.4  37389194  https://earthquake.usgs.gov/fdsnws/event/1/que...   
3   NaN  NaN  37389186  https://earthquake.usgs.gov/fdsnws/event/1/que...   
4   NaN  NaN  73096941  https://earthquake.usgs.gov/fdsnws/event/1/que...   

       dmin  felt    gap           ids   mag magType  ...     status  \
0  0.008693   NaN   85.0  ,ci37389218,  1.35      ml  ...  automatic   
1  0.020030   NaN   79.0  ,ci37389202,  1.29      ml  ...  automatic   
2  0.021370  28.0   21.0  ,ci37389194,  3.42      ml  ...  automatic   
3  0.026180   NaN   39.0  ,ci37389186,  0.44      ml  ...  automatic   
4  0.077990   NaN  192.0  ,nc73096941,  2.16      md  ...  automatic   

            time                          title tsunami        type  \
0  1539475168010  M 1.4 - 9km NE of Aguanga, CA       0  earthquake   
1  1539475129610  M 1.3 - 9km NE of Aguanga, CA       0  earthquake   
2  1539475062610  M 3.4 - 8km NE of Aguanga, CA       0  earthquake   
3  1539474978070  M 0.4 - 9km NE of Aguanga, CA       0  earthquake   
4  1539474716050  M 2.2 - 10km NW of Avenal, CA       0  earthquake   

                                               types     tz        updated  \
0         ,geoserve,nearby-cities,origin,phase-data, -480.0  1539475395144   
1         ,geoserve,nearby-cities,origin,phase-data, -480.0  1539475253925   
2  ,dyfi,focal-mechanism,geoserve,nearby-cities,o... -480.0  1539536756176   
3         ,geoserve,nearby-cities,origin,phase-data, -480.0  1539475196167   
4  ,geoserve,nearby-cities,origin,phase-data,scit... -480.0  1539477547926   

                                                 url parsed_place  
0  https://earthquake.usgs.gov/earthquakes/eventp...   California  
1  https://earthquake.usgs.gov/earthquakes/eventp...   California  
2  https://earthquake.usgs.gov/earthquakes/eventp...   California  
3  https://earthquake.usgs.gov/earthquakes/eventp...   California  
4  https://earthquake.usgs.gov/earthquakes/eventp...   California  

[5 rows x 27 columns]

Here, 

**What was done  :** The CSV file "parsed.cv" is loaded into a pandas dataframe.

**Why            :**  pd.read_csv() is used to import CSV files for analysis using pandas.

**Interpretation :** The dataset is now ready for filtering, calculation and analysis.

## 1.  Percentile Calculation

In [40]:
# Filter earthquakes in Japan with magType 'mb' 
japan_eqk = parsed_df[parsed_df["place"].str.contains("Japan") & (parsed_df["magType"] == "mb")]

percentile_95 = np.percentile(japan_eqk["mag"], 95)

percentile_95


np.float64(4.9)

Here, 

**What was done  :**  We filtered the dataset to select only earthquakes that occurred in Japan and have magType = 'mb'. Then, we calculated the 95th  
                  percentile of their magnitude values.

**Why            :**  Boolean indexing (DataFrame[condition]) is used to filter data based on multiple conditions.  
                  np.percentile() from NumPy calculates the percentile of numeric data efficiently.

**Interpretation :** The result tells us the magnitude below which 95% of the Japanese earthquakes (magType='mb') fall.  
                 The result is 4.9, it means that 95% of these earthquakes have a magnitude less than or equal to 4.9, and only the strongest 5% exceed this value.

##  2.  Tsunami Percentage (Indonesia)

In [41]:
# Select earthquakes in Indonesia
indonesia_eqk = parsed_df[parsed_df["place"].str.contains("Indonesia")]

# Count how many of these caused a tsunami
tsunami_count = indonesia_eqk["tsunami"].sum()  

# Count total earthquakes in Indonesia
total_eqk = len(indonesia_eqk)

# Calculate percentage
tsunami_percentage = (tsunami_count / total_eqk) * 100

round(tsunami_percentage,2)

np.float64(23.13)

Here,

**What was done :**  We filtered the dataset to select only earthquakes that occurred in Indonesia. Then, we counted how many of these earthquakes were 
                 associated with a tsunami (tsunami == 1) and calculated the percentage out of all Indonesia earthquakes.

**Why           :**  Boolean indexing (DataFrame[condition]) is used to filter earthquakes in Indonesia.  
                  Selecting the 'tsunami' column and using .sum() counts all earthquakes where a tsunami occurred (because tsunami = 1).  
                  Dividing by the total number of Indonesia earthquakes gives the percentage of earthquakes that caused tsunamis.

**Interpretation :** The result (e.g., 23.13%) tells us that approximately 23% of Indonesia earthquakes triggered a tsunami, while the rest did not. This 
                 helps understand how often earthquakes in Indonesia are associated with tsunamis.

## 3.  Summary Statistics (Nevada)

In [42]:
# Select earthquakes in Nevada
nevada_eqk = parsed_df[parsed_df['place'].str.contains("Nevada")]

# Generate summary statistics for magnitude
nevada_summary = nevada_eqk['mag'].describe()

nevada_summary

count    677.000000
mean       0.491728
std        0.689560
min       -0.500000
25%       -0.100000
50%        0.400000
75%        0.900000
max        2.900000
Name: mag, dtype: float64

Here, 

**What was done  :** Filtered the dataset to select earthquakes that occurred in Nevada and generated summary statistics for their magnitudes.

**Why            :** Boolean indexing was used to filter rows where the 'place' contains "Nevada".
                 The .describe() function in pandas was used to generate summary statistics such as count, mean, standard deviation, minimum, quartiles, and maximum.
                 
**Interpretation :** Nevada has 677 recorded earthquakes. Most are very small in magnitude, with an average around 0.5 and maximum 2.9. The quartiles show  
                 that the majority are less than 1. Overall, earthquakes in Nevada are mostly minor.

## 4.  Ring of Fire Classification

In [43]:
# List of ring of Fire Locations
ring_locations = [
    "Bolivia", "Chile", "Ecuador", "Peru", "Costa Rica",
    "Guatemala", "Mexico",  # careful about New Mexico
    "Japan", "Philippines", "Indonesia", "New Zealand", "Antarctic",
    "Canada", "Fiji", "Alaska", "Washington", "California",
    "Russia", "Taiwan", "Tonga", "Kermadec Islands"
]

# Create new column
parsed_df["ring_of_fire"] = parsed_df['place'].apply(lambda x: any(place in x for place in ring_locations))

# Exclude New Maxico
parsed_df.loc[parsed_df['place'].str.contains("New Mexico"), "ring_of_fire"] = False

parsed_df.iloc[5:10]

alert  cdi        code                                             detail  \
5   NaN  NaN  2018286011  https://earthquake.usgs.gov/fdsnws/event/1/que...   
6   NaN  NaN    20280432  https://earthquake.usgs.gov/fdsnws/event/1/que...   
7   NaN  NaN    73096936  https://earthquake.usgs.gov/fdsnws/event/1/que...   
8   NaN  NaN    73096931  https://earthquake.usgs.gov/fdsnws/event/1/que...   
9   NaN  NaN    1000hbtn  https://earthquake.usgs.gov/fdsnws/event/1/que...   

       dmin  felt    gap             ids   mag magType  ...           time  \
5  0.437300   NaN  158.0  ,pr2018286011,  2.61      md  ...  1539473686440   
6       NaN   NaN    NaN    ,ak20280432,  1.70      ml  ...  1539473176017   
7  0.016220   NaN   83.0    ,nc73096936,  1.13      md  ...  1539473060280   
8  0.009138   NaN   52.0    ,nc73096931,  0.92      md  ...  1539473042310   
9  3.191000   NaN   37.0    ,us1000hbtn,  4.70      mb  ...  1539472814760   

                                               title  tsunami        type  \
5  M 2.6 - 55km ESE of Punta Cana, Dominican Repu...        0  earthquake   
6               M 1.7 - 105km W of Talkeetna, Alaska        0  earthquake   
7                   M 1.1 - 10km NW of Parkfield, CA        0  earthquake   
8                  M 0.9 - 6km NW of The Geysers, CA        0  earthquake   
9            M 4.7 - 219km SSE of Saparua, Indonesia        0  earthquake   

                                               types     tz        updated  \
5                       ,geoserve,origin,phase-data, -300.0  1539500579236   
6                                  ,geoserve,origin, -540.0  1539473596465   
7  ,geoserve,nearby-cities,origin,phase-data,scit... -480.0  1539476642808   
8  ,geoserve,nearby-cities,origin,phase-data,scit... -480.0  1539475027632   
9                       ,geoserve,origin,phase-data,  540.0  1539473712040   

                                                 url        parsed_place  \
5  https://earthquake.usgs.gov/earthquakes/eventp...  Dominican Republic   
6  https://earthquake.usgs.gov/earthquakes/eventp...              Alaska   
7  https://earthquake.usgs.gov/earthquakes/eventp...          California   
8  https://earthquake.usgs.gov/earthquakes/eventp...          California   
9  https://earthquake.usgs.gov/earthquakes/eventp...           Indonesia   

  ring_of_fire  
5        False  
6         True  
7        False  
8        False  
9         True  

[5 rows x 28 columns]

Here, 

**What was done  :** Added a new column 'ring_of_fire' that is True if the earthquake occurred in a Ring of Fire location and False otherwise.
                 Excluded New Mexico manually.

**Why            :**  The .apply() function with a lambda was used to check each row individually for Ring of Fire locations.  
                  'any()' checks if any location in the list matches the place.  
                  'loc[]' was used to update rows that contain "New Mexico".  
                  'iloc[]' was used to quickly view specific rows to verify results.

**Interpretation :** Rows with True indicate earthquakes in Ring of Fire regions, while False indicates outside. 

## 5. Earthquake Counts (Inside vs Outside Ring of Fire)

In [44]:
# Count earthquakes inside the Ring of Fire
inside_count = parsed_df["ring_of_fire"].sum() # True counts as 1

# Count earthquakes outside the Ring of Fire
outside_count = len(parsed_df) - inside_count

inside_count , outside_count

(np.int64(4426), np.int64(4906))

Here, 

**What was done  :** Calculated the number of earthquakes that occurred inside and outside the Ring of Fire using the 'ring_of_fire' column.

**Why            :**  .sum() counts the number of True values (inside Ring of Fire).  
                  ' len(parsed_df) - inside_count ' gives the number of False values (outside Ring of Fire).

**Interpretation :** Out of all earthquakes in the dataset, 4426 occurred in Ring of Fire locations, while 4906 occurred outside.  
                 This shows that roughly half of the earthquakes are concentrated in the Ring of Fire, highlighting the region's high seismic activity.

## 6.  Tsunami Count Along the Ring of Fire 

In [45]:
# Select rows with earthquakes in Ring of Fire
RoF_eqk = parsed_df[parsed_df["ring_of_fire"] == True]

# Count how many of these caused a tsunami
tsunami_count_ROF = RoF_eqk["tsunami"].sum()

tsunami_count_ROF

np.int64(43)

Here, 

**What was done  :**  Counted how many earthquakes in Ring of Fire locations were associated with a tsunami.

**Why            :**  Boolean indexing was used to select only rows with 'ring_of_fire = True'.  
                The 'sum()' function was applied to the 'tsunami' column (1 = tsunami, 0 = no tsunami) to get the total count.

**Interpretation :**  There were 43 earthquakes in Ring of Fire regions that caused tsunamis.  
                  This shows that while many earthquakes occur in the Ring of Fire, only a small fraction trigger tsunamis.


# PART - II

##  1.  Filtering (earthquakes.csv)

In [46]:
# Load the earthquakes dataset
earthquake_df = pd.read_csv("CSV_Files/earthquakes.csv")

#Filter earthquakes in Japan with magType 'mb' and magnitude >= 4.9
japan_eqk_filtered = earthquake_df[(earthquake_df['place'].str.contains("Japan", na=False)) & (earthquake_df['magType'] == 'mb') & (earthquake_df['mag'] >= 4.9)]

japan_eqk_filtered.head()

mag magType           time                         place  tsunami  \
1563  4.9      mb  1538977532250  293km ESE of Iwo Jima, Japan        0   
2576  5.4      mb  1538697528010    37km E of Tomakomai, Japan        0   
3072  4.9      mb  1538579732490     15km ENE of Hasaki, Japan        0   
3632  4.9      mb  1538450871260    53km ESE of Hitachi, Japan        0   

     parsed_place  
1563        Japan  
2576        Japan  
3072        Japan  
3632        Japan

Here, 

**What was done  :** Filtered earthquakes that occurred in Japan with magType 'mb' and magnitude ≥ 4.9.

**Why            :**  Boolean indexing was used to combine multiple conditions:  
                - 'str.contains("Japan", na=False)' selects rows where 'place' contains "Japan".  
                - '(magType == 'mb')' selects earthquakes with the magnitude type 'mb'.  
                - '(mag >= 4.9)' selects only stronger earthquakes.

**Interpretation :** The filtered DataFrame shows only significant 'mb' earthquakes in Japan.  
                 This helps focus on stronger earthquakes that may be more important for analysis or safety studies.

##  2.  Binning + Counts (earthquakes.csv)

In [47]:
# Filter earthquakes with magType 'ml'
ml_eqk = earthquake_df[earthquake_df["magType"] == 'ml']

# Define bins for each whole number 0 to 10
bins = list(range(0,11)) # [0,1,2,...,10]

# Use pd.cut to put magnitudes into bin
ml_eqk["mag_bin"] = pd.cut(ml_eqk['mag'], bins = bins, right= False)  # right=False means interval is [a,b)

# Count how many earthquakes fall into each bins
bins_count = ml_eqk["mag_bin"].value_counts().sort_index()

bins_count

mag_bin
[0, 1)     2072
[1, 2)     3126
[2, 3)      985
[3, 4)      153
[4, 5)        6
[5, 6)        2
[6, 7)        0
[7, 8)        0
[8, 9)        0
[9, 10)       0
Name: count, dtype: int64

Here, 

**What was done  :** Grouped earthquakes with magType 'ml' into magnitude bins for each whole number (0–1, 1–2, …) and counted how many earthquakes fall into each bin.

**Why            :**  - 'pd.cut()' was used to assign each earthquake to a bin based on its magnitude.  
                  - 'value_counts()' counts how many earthquakes fall into each bin.  
                  - 'sort_index()' ensures bins are displayed in order from smallest to largest.

**Interpretation :**  Most 'ml' earthquakes are small. 0–1 and 1–2 magnitude bins have the highest counts, while strong earthquakes are very rare.

## 3.  Groupby + Resample + Aggregation (faang.csv)

In [48]:
# Load the faang dataset
faang_df = pd.read_csv("CSV_Files/faang.csv")

# Ensure 'date' is datetime type
faang_df["date"] = pd.to_datetime(faang_df["date"])

# Set 'date' as the index (needed for resampling)
faang_df.set_index("date", inplace=True)

# Group by ticker, resample monthly, and compute aggregations
monthly_agg = faang_df.groupby("ticker").resample("ME").agg(
    {
        "open":"mean",
        "high":"max",
        "low":"min",
        "close":"mean",
        "volume":"sum"
    }
)

monthly_agg.head()

open       high        low      close      volume
ticker date                                                              
AAPL   2018-01-31  43.505357  45.025002  41.174999  43.501309  2638717600
       2018-02-28  41.819079  45.154999  37.560001  41.909737  3711577200
       2018-03-31  43.761786  45.875000  41.235001  43.624048  2854910800
       2018-04-30  42.441310  44.735001  40.157501  42.458572  2664617200
       2018-05-31  46.239091  47.592499  41.317501  46.384205  2483905200

Here, 

**What was done :**  Grouped FAANG data by ticker, resampled to month-end, and calculated: mean open, max high, min low, mean close, sum volume.  

**Why :**              'groupby' separates stocks, 'resample("ME")' aggregates monthly, '.agg()' computes multiple stats at once. Index must be datetime for resampling.  

**Interpretation :** This provides a 'monthly summary' for each FAANG stock, showing trends in price and total trading volume. It helps analyze stock performance over time in a concise format.

## 4. Crosstab with Aggregation (earthquakes.csv)

In [49]:
# Build crosstab showing max magnitude for each tsunami vs magType combination

tsunami_mag_crosstab = earthquake_df.pivot_table(
    index="tsunami",    # Rows: tsunami occurrence
    columns="magType",  # Columns: magnitude type
    values="mag",       # What we are measuring
    aggfunc="max"       # Take the maximum value instead of count
)

tsunami_mag_crosstab

magType   mb  mb_lg    md   mh   ml  ms_20    mw  mwb  mwr  mww
tsunami                                                        
0        5.6    3.5  4.11  1.1  4.2    NaN  3.83  5.8  4.8  6.0
1        6.1    NaN   NaN  NaN  5.1    5.7  4.41  NaN  NaN  7.5

Here, 

**What was done  :**  Created a table showing the 'largest earthquake magnitude' for each combination of tsunami occurrence and magnitude type. 

**Why            :** Used 'pivot_table()' to summarize data and 'aggfunc="max" ' to get the maximum magnitude instead of counts.  

**Interpretation :** The table shows the largest earthquake magnitudes for each magnitude type, separated by whether a tsunami occurred (1) or not (0). For example, the largest 
                     mb earthquake without a tsunami was 5.6, while the largest mb earthquake with a tsunami was 6.1. The largest mww earthquake that caused a tsunami was 7.5. Any NaN values indicate that no earthquakes of that type occurred in that category.

## 5.  Z-scores with apply (faang.csv)

In [50]:
# Select only Netflix data from faang dataset
netflix_df = faang_df[faang_df["ticker"] == "NFLX"]

# Calculate Z-scores for numeric columns only
netflix_zscores = netflix_df.select_dtypes(include="number").apply(lambda x: (x - x.mean()) / x.std())

netflix_zscores

high       low      open     close    volume
date                                                        
2018-01-02 -2.515825 -2.410229 -2.500752 -2.416646 -0.088638
2018-01-03 -2.422985 -2.285796 -2.380291 -2.335287 -0.507472
2018-01-04 -2.405883 -2.234631 -2.296272 -2.323431 -0.959154
2018-01-05 -2.345415 -2.202090 -2.275014 -2.234304 -0.782205
2018-01-08 -2.294923 -2.143761 -2.218933 -2.192194 -1.038390
...              ...       ...       ...       ...       ...
2018-12-24 -1.518204 -1.627199 -1.571478 -1.745948 -0.338881
2018-12-26 -1.439819 -1.677341 -1.735063 -1.341403  0.517143
2018-12-27 -1.417627 -1.495807 -1.407286 -1.302563  0.134982
2018-12-28 -1.288954 -1.297287 -1.248762 -1.292138 -0.084071
2018-12-31 -1.122209 -1.088533 -1.203817 -1.055421  0.359553

[251 rows x 5 columns]

Here, 

**What was done  :** The numeric columns ('open', 'high', 'low', 'close', 'volume') have been standardized to Z-scores. 

**Why            :** - 'select_dtypes(include="number")' was used to select only numeric columns, so Z-scores are calculated only where it makes sense.  
                     - 'apply()' was used to apply the Z-score formula column by column.  
                     - '(x - x.mean()) / x.std()' standardizes each column relative to its own mean and standard deviation.

**Interpretation :** The table shows how each value compares to its column average in standard deviations. Values near 0 are close to the mean, positive values are above 
                      average, and negative values are below average. This helps identify unusually high or low stock values relative to typical behavior.

## 6.  Add Event Descriptions (faang.csv)

In [51]:
# Reset index to make date to column
faang_df = faang_df.reset_index()

# Convert date to column to datetime
faang_df['date'] = pd.to_datetime(faang_df['date'])

# Set MultiIndex (date + ticker)
faang_df.set_index(["date", "ticker"], inplace=True)

# Create a DataFrame with columns ticker, date, and event 
event_df = pd.DataFrame({
    "ticker":['FB','FB','FB'],
    "date": ['2018-07-25', '2018-03-19', '2018-03-20'],
    "event": ['Disappointing user growth announced after close.', 'Cambridge Analytica story', 'FTC investigation']
})

# Convert date column to datetime
event_df['date'] = pd.to_datetime(event_df['date'])

# Set a multi-index on date and ticker
event_df.set_index(['date', 'ticker'], inplace=True)

# Merge with the FAANG data using outer join
event_faang_merged = faang_df.merge(event_df, how="outer",left_index=True, right_index=True)

event_faang_merged.loc[event_faang_merged['event'].notna()]

high         low        open       close     volume  \
date       ticker                                                              
2018-03-19 FB      177.169998  170.059998  177.009995  172.559998   88140100   
2018-03-20 FB      170.199997  161.949997  167.470001  168.149994  129851800   
2018-07-25 FB      218.619995  214.270004  215.720001  217.500000   58954200   

                                                              event  
date       ticker                                                    
2018-03-19 FB                             Cambridge Analytica story  
2018-03-20 FB                                     FTC investigation  
2018-07-25 FB      Disappointing user growth announced after close.

Here, 

**What was done  :** Added specific Facebook events on given dates to the FAANG dataset by creating a separate DataFrame and merging it.  

**Why            :** - 'pd.to_datetime()' was used to ensure the event dates have the same type as the FAANG data ('datetime64'), which is required for merging.  
                     - 'set_index(['date', 'ticker'])' aligns the event DataFrame with FAANG for merging.  
                     - 'merge(..., how="outer")' keeps all FAANG stock data while adding event information where available.  

**Interpretation :** Only FB rows on the event dates show the event description; all other rows have NaN. This links specific stock movements to real-world events.

## 7.  Transform to Index Values (faang.csv)

In [52]:
# Reset index so ticker and date are columns
faang_df = faang_df.reset_index()

#Drop extra index columns if they exist
faang_df = faang_df.drop(columns=[col for col in ['level_0', 'index'] if col in faang_df.columns])

# Select numeric columns
numeric_cols = faang_df.select_dtypes(include='number').columns

# Copy dataframe
faang_indexed = faang_df.copy()

# Convert numeric values relative to first date of each ticker
faang_indexed[numeric_cols] = faang_indexed.groupby('ticker')[numeric_cols].transform(lambda x: x / x.iloc[0])

# Keep the ticker and data column for reference
faang_indexed = faang_indexed[['date','ticker'] + list(numeric_cols)]

faang_indexed.head()

date ticker      high       low      open     close    volume
0 2018-01-02     FB  1.000000  1.000000  1.000000  1.000000  1.000000
1 2018-01-03     FB  1.017623  1.021290  1.023638  1.017914  0.930294
2 2018-01-04     FB  1.025498  1.036891  1.040635  1.016040  0.764708
3 2018-01-05     FB  1.029298  1.041566  1.044518  1.029931  0.747828
4 2018-01-08     FB  1.040313  1.049451  1.053579  1.037813  0.991340

Here,

**What was done  :** : Converted all numeric stock values (high, low, open, close, volume) to be relative to the first date for each ticker (first-date value = 1) using  
                       groupby().transform()

**Why            :**  reset_index() ensures date and ticker are columns for grouping.
                      groupby('ticker') separates data by company.
                      transform(lambda x: x / x.iloc[0]) scales all numeric values relative to the first row per ticker.

**Interpretation :**  All numeric values now show relative change from the first available date per ticker. For example, a value of 1.05 means a 5% increase compared to the first
                      date.